### PR TITLE
[APIM] Add product / productapi / namedValue entity commands && migrate sdk version

### PR DESCRIPTION
**Pull request by RupengLiu**
**Description**  

**Testing Guide**  

**History Notes**  

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.  
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
APIM

File: src/azure-cli/azure/cli/command_modules/apim/_params.py [hunk 1]
```diff
@@ -133,3 +133,82 @@ def load_arguments(self, _):
...
+        c.argument('subscriptions_limit', help="Whether the number of subscriptions a user can have to this product at the same time. Set to null or omit to allow unlimited per user subscriptions. Can be present only if subscriptionRequired property is present and has a value of false.")
+        c.argument('state', help="whether product is published or not. Published products are discoverable by users of developer portal. Non published products are visible only to administrators. Default state of Product is notPublished. Possible values include: 'notPublished', 'published'")
+
+    with self.argument_context('apim product delete') as c:
+        c.argument('service_name', options_list=['--service-name', '-n'], help="The name of the api management service instance", id_part=None)
+        c.argument('product_id', help="Product identifier. Must be unique in the current API Management service instance.")
+        c.argument('delete_subscriptions', help="Delete existing subscriptions associated with the product or not.")
+        c.argument('if_match', help='ETag of the Entity.')
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/apim/_params.py:189]: is the help message correct?

File: src/azure-cli/azure/cli/command_modules/apim/_help.py [hunk 2]
```diff
@@ -147,3 +162,144 @@
...
+  - name: Check if the API is associated with the Product.
+    text: |-
+        az apim productapi check --resource-group MyResourceGroup  --service-name MyServiceName  --product-id MyProductID --api-id MyAPIID
+"""
+
+helps['apim productapi add'] = """
+type: command
+short-summary: Adds an API to the specified product.
+examples:
+  - name: Adds an API to the specified product.
```
[the remark this answers is not in the document]
Comment (RupengLiu) [reply src/azure-cli/azure/cli/command_modules/apim/_help.py:188]: sounds good

**Comment by ThorstenHans** [comment]
@RupengLiu are you actively working on the feedback provided by the team? Do you have a rough time schedule for finishing the PR? 

Or can I submit PRs helping to your fork, helping to incorporate the feedback :D

File: src/azure-cli/azure/cli/command_modules/apim/_params.py (hunk 1 above)
Comment (RupengLiu) [reply src/azure-cli/azure/cli/command_modules/apim/_params.py:189]: yep~

**Comment by RupengLiu** [comment]
@ThorstenHans thanks for helping, I was just working on other stuffs, will update the PR today~

**Comment by RupengLiu** [comment]
@MyronFanQiu Hi Fan, for the linter error, I got "option_length_too_long" for "subscription-required", but in the previous, we still have this "option", how do I suppress this linter error, do I have to specially deal with this ?

**Comment by MyronFanQiu** [comment]
> @MyronFanQiu Hi Fan, for the linter error, I got "option_length_too_long" for "subscription-required", but in the previous, we still have this "option", how do I suppress this linter error, do I have to specially deal with this ?

You just need to provide one extra short option.

File: src/azure-cli/azure/cli/command_modules/apim/commands.py [hunk 3]
```diff
@@ -39,3 +49,23 @@ def load_command_table(self, _):
...
+
+    with self.command_group('apim product api', api_sdk, is_preview=True) as g:
+        g.custom_command('list', 'list_product_api')
+        g.custom_command('check', 'check_product_exists')
+        g.custom_command('add', 'add_product_api')
+        g.custom_command('delete', 'delete_product_api')
+
+    with self.command_group('apim product', product_sdk, is_preview=True) as g:
+        g.custom_command('list', 'list_products')
+        g.custom_command('show', 'show_product')
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/apim/commands.py:61]: please use `costom_show_command`

**Comment by MyronFanQiu** [comment]
@RupengLiu Please do the changes as we talked offline and I can help you merge this PR.

**Comment by MyronFanQiu** [comment]
This PR is too large. Better not add more commands into it. The review process will be longer.

**Comment by RupengLiu** [comment]
@MyronFanQiu any update for this PR~

File: src/azure-cli/azure/cli/command_modules/apim/_params.py [hunk 4]
```diff
@@ -133,3 +134,100 @@ def load_arguments(self, _):
...
+        c.argument('subscriptions_limit', help="Whether the number of subscriptions a user can have to this product at the same time. Set to null or omit to allow unlimited per user subscriptions. Can be present only if subscriptionRequired property is present and has a value of false.")
+        c.argument('state', help="whether product is published or not. Published products are discoverable by users of developer portal. Non published products are visible only to administrators. Default state of Product is notPublished. Possible values include: 'notPublished', 'published'")
+
+    with self.argument_context('apim product update') as c:
+        c.argument('service_name', options_list=['--service-name', '-n'], help="The name of the api management service instance", id_part=None)
+        c.argument('product_id', help="Product identifier. Must be unique in the current API Management service instance.")
+        c.argument('product_name', help="Required. Product name.")
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/apim/_params.py:178]: Please remove reduntdant code to avoid such issue.

**Review by MyronFanQiu — COMMENTED** [review]
Could you add test for these four commands?
apim apply-network-updates doesn't have test
apim wait doesn't have test
apim api wait doesn't have test
apim nv show-secret doesn't have test

**Review by MyronFanQiu — APPROVED** [review]

**Pull request merged**